libraries/networking/src/AssetClient.cpp

```diff
@@ -0,0 +1,208 @@
+    auto assetHash = QString(packet->read(HASH_HEX_LENGTH));
+    
+    AssetServerError error;
+    packet->readPrimitive(&error);
+
+    AssetInfo info;
+
+    if (!error) {
+        packet->readPrimitive(&info.size);
+    }
+
+    if (_pendingInfoRequests.contains(messageID)) {
+        auto callback = _pendingInfoRequests.take(messageID);
+        callback(error != NO_ERROR, info);
+    }
+}
+
+void AssetClient::handleAssetGetReply(QSharedPointer<NLPacketList> packetList, SharedNodePointer senderNode) {
+    QByteArray data = packetList->getMessage();
+    QBuffer packet { &data };
+    packet.open(QIODevice::ReadOnly);
+
+    auto assetHash = packet.read(HASH_HEX_LENGTH);
+    qDebug() << "Got reply for asset: " << assetHash;
+
+    MessageID messageID;
+    packet.read(reinterpret_cast<char*>(&messageID), sizeof(messageID));
+
+    AssetServerError error;
+    packet.read(reinterpret_cast<char*>(&error), sizeof(AssetServerError));
+    QByteArray assetData;
+
+    if (!error) {
+        DataOffset length;
+        packet.read(reinterpret_cast<char*>(&length), sizeof(DataOffset));
+        data = packet.read(length);
+    } else {
+        qDebug() << "Failure getting asset: " << error;
+    }
+
+    if (_pendingRequests.contains(messageID)) {
+        auto callback = _pendingRequests.take(messageID);
+        callback(!error, data);
+    }
+}
+
+bool AssetClient::uploadAsset(QByteArray data, QString extension, UploadResultCallback callback) {
+    auto nodeList = DependencyManager::get<NodeList>();
+    SharedNodePointer assetServer = nodeList->soloNodeOfType(NodeType::AssetServer);
+    if (assetServer) {
+        auto packetList = std::unique_ptr<NLPacketList>(new NLPacketList(PacketType::AssetUpload, QByteArray(), true, true));
+
+        auto messageID = _currentID++;
+        packetList->writePrimitive(messageID);
+
+        packetList->writePrimitive(static_cast<uint8_t>(extension.length()));
+        packetList->write(extension.toLatin1().constData(), extension.length());
+
+        qDebug() << "Extension length: " << extension.length();
+        qDebug() << "Extension: " << extension;
+
+        uint64_t size = data.length();
+        packetList->writePrimitive(size);
+        packetList->write(data.constData(), size);
+
+        nodeList->sendPacketList(std::move(packetList), *assetServer);
+
+        _pendingUploads[messageID] = callback;
+
+        return true;
+    }
+    return false;
+}
+
+void AssetClient::handleAssetUploadReply(QSharedPointer<NLPacket> packet, SharedNodePointer senderNode) {
+    qDebug() << "Got asset upload reply";
+    MessageID messageID;
+    packet->readPrimitive(&messageID);
+
+    bool success;
+    packet->readPrimitive(&success);
+
+    QString hashString { "" };
+
+    if (success) {
+        auto hashData = packet->read(HASH_HEX_LENGTH);
+
+        hashString = QString(hashData);
+
+        qDebug() << "Hash: " << hashString;
+    } else {
+        qDebug() << "Error uploading file";
+    }
```

Comment:
I think it would be cleaner in the log remove the top debug and add some context to the two debugs just above this comment. Something like:

Successfully uploaded asset to asset-server - SHA256 hash is $HASH
Error uploading file to asset-server.